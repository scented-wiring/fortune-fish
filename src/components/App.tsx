import { useEffect, useState } from "react";
import $ from "jquery";
import hand from "../images/hand.png";
import head from "../images/head.png";
import body from "../images/body.png";
import tail from "../images/tail.png";

import "../styles/App.css";

const App = () => {
  const [gameStatus, setGameStatus] = useState("prefish:initial");
  const [notReadyClick, setNotReadyClick] = useState(0);

  let fortune, message, buttons;

  const handleFortune = () => {
    setGameStatus("pause");
    const num = Math.floor(Math.random() * 7) + 1;
    if (num === 7) {
      $("#head").attr("id", "head-move");
      setTimeout(() => {
        setGameStatus("fortune:Jealous");
      }, 1500);
    }
    if (num === 6) {
      $("#tail").attr("id", "tail-move");
      setTimeout(() => {
        setGameStatus("fortune:Indifferent");
      }, 1500);
    }
    if (num === 5) {
      $("#tail").attr("id", "tail-move");
      $("#head").attr("id", "head-move");
      setTimeout(() => {
        setGameStatus("fortune:In Love");
      }, 1500);
    }
    if (num === 4) {
      $("#tail").attr("id", "tail-curl");
      $("#head").attr("id", "head-curl");
      setTimeout(() => {
        setGameStatus("fortune:Fickle");
      }, 1500);
    }
    if (num === 3) {
      $("#fish").attr("id", "jump");
      setTimeout(() => {
        setGameStatus("fortune:False");
      }, 1500);
    }
    if (num === 2)
      setTimeout(() => {
        setGameStatus("fortune:Tired");
      }, 1500);
    if (num === 1) {
      $("#fish").attr("id", "jump-nose");
      setTimeout(() => {
        setGameStatus("fortune:Passionate");
      }, 1500);
    }
  };

  const handleNotReadyClick = () => {
    setGameStatus("prefish:not ready");
    setNotReadyClick(notReadyClick + 1);
  };

  useEffect(() => {
    //spotlight fades in on load
    $("#spotlight-hide").attr("id", "spotlight");
  });

  if (gameStatus !== "prefish:initial") {
    //hand slides in after initial screen
    $("#hand-hide").attr("id", "hand");
  }

  if (!gameStatus.includes("prefish:")) {
    //fish floats down on cue
    $("#fish-hide").attr("id", "fish");
  }

  if (gameStatus.includes("fortune:")) {
    //set fortune and "try again" button at appropriate time
    fortune = gameStatus.replace("fortune:", "");
    buttons = (
      <div>
        <button onClick={() => setGameStatus("try again")}>Try Again</button>
      </div>
    );
  }

  if (
    gameStatus === "prefish:palm presented" ||
    gameStatus === "prefish:not ready"
  ) {
    //renders ready options
    buttons = (
      <div>
        <button onClick={() => setGameStatus("ready")}>Yes</button>
        <button onClick={handleNotReadyClick}>No</button>
      </div>
    );
  }

  if (gameStatus === "prefish:not ready" && notReadyClick === 3) {
    //hand slides out after three "not ready" presses
    $("#hand").attr("id", "hand-hide");
  }

  if (gameStatus === "prefish:initial") {
    message = "Present your palm, my child.";
    buttons = (
      <div>
        <button onClick={() => setGameStatus("prefish:palm presented")}>
          OK
        </button>
      </div>
    );
  } else if (gameStatus === "prefish:palm presented") {
    message = "Are you ready to hear your fortune?";
  } else if (gameStatus === "prefish:not ready" && notReadyClick === 1) {
    message = "How about now?";
  } else if (gameStatus === "prefish:not ready" && notReadyClick === 2) {
    message = "...Now?";
  } else if (gameStatus === "prefish:not ready" && notReadyClick === 3) {
    message = "Right, just forget it then";
    buttons = null;
  } else if (gameStatus === "ready") {
    message = "He comes...";
    setTimeout(() => {
      setGameStatus("waiting");
    }, 3000);
  } else if (gameStatus === "waiting") {
    message = null;
    buttons = (
      <div>
        <button onClick={handleFortune}>Well?</button>
      </div>
    );
  } else if (gameStatus === "pause") {
    message = null;
    buttons = null;
  } else if (gameStatus === "fortune:Jealous") {
    message =
      "See how the head moves? The fortune fish has determined you to be a jealous person. A most ugly trait.";
  } else if (gameStatus === "fortune:Indifferent") {
    message =
      "A wagging tail... The fortune fish has determined that you are indifferent. Show a little enthusiasm once in a while.";
  } else if (gameStatus === "fortune:In Love") {
    message =
      "Ah, he is moving both head and tail. The fortune fish has determined that you are in love. Good for you.";
  } else if (gameStatus === "fortune:Fickle") {
    message =
      "See how he curls? He has determined that you are a fickle person. You should have courage in your convictions.";
  } else if (gameStatus === "fortune:False") {
    message =
      "A sick 180 flip! What a sight! But this means the fortune fish has determined you to be false. Try sincerity some time.";
  } else if (gameStatus === "fortune:Tired") {
    message =
      "The fish is motionless! This means that you are tired, or possibly dead. Have you checked your pulse recently?";
  } else if (gameStatus === "fortune:Passionate") {
    message =
      "Astounding! This balancing act means that you are a passionate person. What are you doing later?";
  } else if (gameStatus === "try again") {
    message = "You can't have another go. The fish has spoken. Accept it.";
    buttons = null;
  }
  return (
    <div id="App">
      <h1>FORTUNE FISH</h1>
      <div id="spotlight-hide" />
      <div id="fish-hide">
        <img src={head} alt="fish head" id="head" />
        <img src={body} alt="fish body" id="body" />
        <img src={tail} alt="fish tail" id="tail" />
      </div>
      <img id="hand-hide" src={hand} alt="hand" width="200px" />
      <div id="ui">
        <p id={!fortune ? "fortune-hide" : "fortune"}>
          {!fortune ? null : fortune.toUpperCase()}
        </p>
        <p id="message">{message}</p>
        <div id="buttons">{buttons}</div>
      </div>
    </div>
  );
};

export default App;
