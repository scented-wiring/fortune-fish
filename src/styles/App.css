#App {
  margin-top: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
}

#spotlight-hide {
  opacity: 0;
  margin: 120px 0 30px 0;
  height: 60px;
  width: 300px;
  border-radius: 50%;
  background-color: black;
  box-shadow: 0 0 20px 20px black;
}

#spotlight {
  opacity: 1;
  margin: 120px 0 30px 0;
  height: 60px;
  width: 300px;
  border-radius: 50%;
  background-color: white;
  box-shadow: 0 0 20px 20px white;
  transition: all 2s ease 0s;
}

#message {
  width: 320px;
  text-align: center;
}

#hand-hide {
  position: absolute;
  margin-top: 250px;
  margin-left: 15px;
}

#hand {
  transition: all 1s ease 0s;
  position: absolute;
  margin-top: 75px;
  margin-left: 15px;
}

#fish-hide {
  opacity: 0;
  position: absolute;
  margin-right: 10px;
}

#fish {
  z-index: 1;
  position: absolute;
  margin-top: 158px;
  margin-right: 18px;
  transition: all 2s ease 0s;
}

#head-move {
  transform-origin: right;
  animation: move 1s infinite;
  animation-timing-function: ease-in-out;
}

#tail-move {
  transform-origin: left;
  animation: move 1s infinite;
  animation-timing-function: ease-in-out;
}

#jump {
  z-index: 1;
  position: absolute;
  margin-top: 155px;
  margin-right: 10px;
  animation: jump 0.5s;
  animation-timing-function: ease-in-out;
  animation-fill-mode: forwards;
}

#jump-nose {
  z-index: 1;
  position: absolute;
  margin-top: 155px;
  margin-right: 10px;
  animation: jump-nose 0.5s;
  animation-timing-function: ease-in-out;
  animation-fill-mode: forwards;
}

#head-curl {
  transform-origin: bottom right;
  animation: head-curl 1s;
  animation-fill-mode: forwards;
}

#tail-curl {
  transform-origin: bottom left;
  animation: tail-curl 1s;
  animation-fill-mode: forwards;
}

#ui {
  z-index: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  background: black;
  height: 200px;
}

@keyframes move {
  0% {
    transform: scaleX(1);
  }
  25% {
    transform: scaleX(1.2);
  }
  50% {
    transform: scaleX(1);
  }
  75% {
    transform: scaleX(0.7);
  }
  100% {
    transform: scaleX(1);
  }
}

@keyframes jump {
  75% {
    transform: rotate(90deg) translateX(-50px);
  }
  100% {
    transform: rotate(180deg);
  }
}

@keyframes jump-nose {
  75% {
    transform: rotate(180deg) translateY(100px);
  }
  100% {
    transform: rotate(270deg) translateX(40px);
  }
}

@keyframes head-curl {
  50% {
    transform: skewY(10deg);
  }
  100% {
    transform: skewY(20deg);
  }
}

@keyframes tail-curl {
  50% {
    transform: skewY(-10deg);
  }
  100% {
    transform: skewY(-20deg);
  }
}
