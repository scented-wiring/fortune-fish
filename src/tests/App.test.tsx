import React from "react";
import { fireEvent, render } from "@testing-library/react";
import App from "../components/App";

describe("App component", () => {
  test("renders correctly", () => {
    const { asFragment } = render(<App />);
    expect(asFragment).toMatchSnapshot();
  });

  test("renders message and hidden hand on initial load", () => {
    const container = render(<App />);
    expect(
      container.getByText("Present your palm, my child.")
    ).toBeInTheDocument();
    expect(document.getElementById("hand-hide")).toBeInTheDocument();
  });

  test("renders new message and hand after OK press", () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    expect(
      container.getByText("Are you ready to hear your fortune?")
    ).toBeInTheDocument();
    expect(document.getElementById("hand")).toBeInTheDocument();
  });

  test("renders message on first 'no' press", () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    fireEvent.click(container.getByText("No"));
    expect(container.getByText("How about now?")).toBeInTheDocument();
  });

  test("renders message on second 'no' press", () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    fireEvent.click(container.getByText("No"));
    fireEvent.click(container.getByText("No"));
    expect(container.getByText("...Now?")).toBeInTheDocument();
  });

  test("renders message and hides hand on third 'no' press", () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    fireEvent.click(container.getByText("No"));
    fireEvent.click(container.getByText("No"));
    fireEvent.click(container.getByText("No"));
    expect(
      container.getByText("Right, just forget it then.")
    ).toBeInTheDocument();
    expect(document.getElementById("hand-hide")).toBeInTheDocument();
  });

  test("renders message and fish on 'yes' press", () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    fireEvent.click(container.getByText("Yes"));
    expect(container.getByText("He comes...")).toBeInTheDocument();
  });

  test("renders button after 3s", async () => {
    const container = render(<App />);
    fireEvent.click(container.getByText("OK"));
    fireEvent.click(container.getByText("Yes"));
    await new Promise((r) => setTimeout(r, 3000));
    expect(container.getByRole("button")).toBeInTheDocument();
  });
});
