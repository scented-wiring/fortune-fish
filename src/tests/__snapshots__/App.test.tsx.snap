// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `[Function]`;

exports[`App component renders message on initial load 1`] = `
<p
  id="message"
>
  Present your palm, my child.
</p>
`;
